File: src/main/java/org/vaadin/bugrap/cdi/MCDIUIProvider.java
package org.vaadin.bugrap.cdi;

import com.vaadin.cdi.CDIUIProvider;

import javax.enterprise.inject.spi.BeanManager;
import javax.enterprise.inject.spi.CDI;

public class MCDIUIProvider extends CDIUIProvider {
    BeanManager beanManager = null;

    @Override
    public BeanManager getBeanManager() {

        if (beanManager == null) {
            beanManager = CDI.current().getBeanManager();
        }
        if (beanManager == null) {
            beanManager = super.getBeanManager();
        }
        return beanManager;
    }
}

File: src/main/java/org/vaadin/bugrap/cdi/jettybootstrap/MCDIUIProvider.kt
package org.vaadin.bugrap.cdi.jettybootstrap

import com.vaadin.cdi.CDIUIProvider
import javax.enterprise.inject.spi.CDI

class MCDIUIProvider : CDIUIProvider() {

  private val manager by lazy { CDI.current().beanManager ?: super.getBeanManager() }

  override fun getBeanManager() = manager
}
